feat(mobile): add deduplication hash to backup headers

jpros opened the pull request:
## Description

This PR introduces a mechanism to prevent duplicated asset uploads to the server. It leverages the existing x-immich-checksum header to send the asset’s SHA-1 checksum for validation before the full upload takes place.

Additionally, this PR implements a single-file checksum bridge between iOS and Android for Flutter, ensuring consistency across platforms.

By preventing unnecessary uploads, especially of large files, this change optimizes bandwidth usage and improves overall efficiency.

**Next Steps**

Since images with different resolutions or compression settings generate distinct SHA-1 checksums, the next step will be to explore a more robust approach to image deduplication.

**Fixes** 
#1553 - Move to client side hashing

## How Has This Been Tested?

I ran the iOS and Android app, tested the backup with new images, tested uploading an image to the server through the web, and then added the same asset to the device. I tried to back up this new file, and it detected the duplicate.



## Checklist:

- [X] I have performed a self-review of my own code
- [N/A] I have made corresponding changes to the documentation if applicable
- [X] I have no unrelated changes in the PR.
- [N/A] I have confirmed that any new dependencies are strictly necessary.
- [N/A] I have written tests for new code (if applicable)
- [X] I have followed naming conventions/patterns in the surrounding code
- [X] All code in `src/services` uses repositories implementations for database calls, filesystem operations, etc.
- [X] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services`)

Commit:
Added hash calculation to backup headers, preventing existing files from being reuploaded.

Commit:
[FIX] Android implementation

Commit:
Checking local asset db for the existing hash. Otherwise, calculate the hash.

bo0tzz commented on mobile/ios/Runner/BackgroundSync/BackgroundServicePlugin.swift:
Please remove the unrelated formatting changes.

jpros replied:
I think the extra spaces are unnecessary. Would you like me to add them back to the file?
(They were automatically removed by Intellij)

bo0tzz commented on mobile/lib/services/backup.service.dart:
```diff
@@ -335,6 +341,19 @@ class BackupService {
             }
           }
 
+          final localAsset = localAssets.firstWhereOrNull(
+            (localAsset) => localAsset.localId == asset.localId,
+          );
```
Looping over all of the assets seems like a bad idea. I think you want [AssetRepository#getDeviceAssetsById](https://github.com/immich-app/immich/blob/f386b4d3771c57560d5d8f973cfb695433703999/mobile/lib/repositories/asset.repository.dart#L154C30-L154C49) instead?

bo0tzz reviewed (changes requested):
Thanks!

Commit:
Added a method on the repository to get an asset by the local ID
Backup service now, instead of loading all local assets, is getting the current asset from the DB using the local ID.

Commit:
Remove fix formatting file

Commit:
Fix backup service format

alextran1502 commented:
Hello, and thank you for the PR. I ran some tests and noticed the feature isn't working as intended. Perhaps it is related to our Dart `http` client.

The duplicated assets are still getting uploaded to the server. 

You can reproduce by connecting to a slower WIFI frequency (2.4GHz) and uploading a larger duplicated file, the expected behavior is that the request would be skipped, but I am still seeing the upload bar progressing

Gabriel-434 commented:
> The duplicated assets are still getting uploaded to the server.

In `backup.service.dart` the `x-immich-checksum` header is not set if the file digest fails, thus uploading the file regardless.

So I think the intended behavior is to gracefully fail and backup the asset.
